README.md
Express Combo
=============

Combo handler for express applications.

[![Build Status](https://travis-ci.org/yahoo/express-combo.png?branch=master)][Build Status]

[Build Status]: https://travis-ci.org/yahoo/express-combo

Goals
-----

 * register and serve static assets from a specific folder
 * register and serve static assets based on an explicit mapping to protect other
   files that should not be exposed
 * serve combo urls based on the static assets that were registered

Installation
------------

Install using npm:

```shell
$ npm install express-combo
```

Usage
-----

```
var express = require('express'),
    app = express(),
    statichandler = require('express-combo');

app.get('/static', statichandler.folder('public', 'full/path/to/folder/'));

app.get('/static', statichandler.map('protected', {
    'something/foo.js": "full/path/to/something/foo.js',
    'bar.js": "full/path/to/something/bar.js'
}));

app.get('/static', statichandler.combine({
    comboBase: '/combo~',
    comboSep: '~'
}));
```

The example above will allow you to access any file within the folder
`full/path/to/folder/` by following the route:

 * http://hostname:port/static/public/any/file.js

In this case, serving files without any protection, which means all files could be
accessed. Under the hood this is equivalent to use `express.static` middleware.
It also expose two files, but this time adding some explicit mapping for them, exposing:

 * http://hostname:port/static/protected/something/foo.js
 * http://hostname:port/static/protected/bar.js

It protects any other file within the those folders. It also provides a nice abstraction
where filenames and paths in the filesystem are not longer relevant when it comes to serve
them, and the mapping has to be explicit.

And last, but not least, it turns on the combo capabilities for all the previous
registered assets, and doing so by defining the separator token. As a result,
a urls like these will be valid:

 * http://hostname:port/static/combo~something/foo.js~bar.js
 * http://hostname:port/static/combo~any/file.js~something/foo.js~bar.js

_Note: the `/static` prefix is optional, and you can remove it all together, and even
use `app.use()` directly for those middleware._

TODO
----

 * support custom filters (regex) to include and/or exclude assets thru `folder` and `map`.
 * css relative path correction when serving thru combo

License
-------

This software is free to use under the Yahoo! Inc. BSD license.
See the [LICENSE file][] for license text and copyright information.

[LICENSE file]: https://github.com/yahoo/express-combo/blob/master/LICENSE.md

Contribute
----------

See the [CONTRIBUTE file][] for info.

[CONTRIBUTE file]: https://github.com/yahoo/express-combo/blob/master/CONTRIBUTE.md


HISTORY.md
Express Combo Change History
============================

0.1.1 (2013-09-16)
------------------

* Relaxing express version to `3.x` now that ynpm registry is stable again.

0.1.1 (2013-09-09)
------------------

* Locking down express to `3.3.x` to bypass issues with ynpm for now.

0.0.1 (2013-03-04)
------------------

* Initial release

tests/fixtures/one.js


// a valid JS file for testing

YUI.add('one-module', function (Y, NAME) {
    Y.One = {
        one: 'one'
    };
});


lib/static.js
/*
 * Copyright (c) 2013, Yahoo! Inc.  All rights reserved.
 * Copyrights licensed under the New BSD License.
 * See the accompanying LICENSE file for terms.
 */

/*jslint node:true, nomen: true */

/**
The `static` middleware provides a set of features
to serve files (static assets), as well as serving
them combined by using the combohandler technique.

@module static
**/

'use strict';

var debug   = require('debug')('express-combo:static'),
    express = require('express'),
    libfs   = require('fs'),
    libpath = require('path'),
    libmime = require('mime'),
    SendStream = require('./send'),

    strategy = require('yui-pathogen-decoder'), // custom combo decoding strategy
    yuicombo = require('yui-combo-middleware'), // custom-combo-related middleware

    // Create a composite middleware out of individual middleware
    decoder = yuicombo.createComposite(
        // decode urls encoded using the pathogen strategy, into module names
        yuicombo.decode(strategy),
        // transform module names into module paths
        yuicombo.path()
    ),

    // if no matching mimeType is found use this
    DEFAULT_MIME_TYPE = 'text/plain; charset=UTF-8';

function dedupe(array) {
    var hash    = {},
        results = [],
        hasOwn  = Object.prototype.hasOwnProperty,
        i,
        item,
        len;

    for (i = 0, len = array.length; i < len; i += 1) {
        item = array[i];

        if (!hasOwn.call(hash, item)) {
            hash[item] = 1;
            results.push(item);
        }
    }

    return results;
}

/**
The `static` provides capabilities to serve static and combined
YUI modules and assets, as well as the ability to map paths and
protect folders by exposing those files that are part of the
mapping configuration only.

    var middleware = require('express-combo');

    // serving static files from a public folder where
    // all files will be exposed.
    app.get('/static/', middleware.folder(__dirname + '/static/', {
        maxAge: 31536000
    });

    // serving static files from a protected folder where
    // no other files in the folder will be exposed except 
    // for the file specified in the map.
    app.get('/other/', middleware.map(__dirname, {
        "assets/something.js": __dirname + "/foo/bar.js"
    });

This component also provide combohandler capabilities, and it is
inspired on the `combohandler` npm pkg, but it focus on the
ability to map paths and define groups for maps.

    // Creates a new express app.
    app.get('/foo', middleware.combine(__dirname, {
        maxAge: 31536000,
        map: {
            yui: {
                path: __dirname + "/yui/build/",
                root: 'yui/'
            },
            app: {
                root: 'app/',
                map: {
                    "assets/something.js": __dirname + "/foo/bar.js"
                }
            }
        }
    });

@class static
@static
@uses *express, *fs, *path, *util
*/
module.exports = {

    /**
    Adds a group of modules into the internal cache structure
    that will be used to map public paths with filesystem paths
    as well as module names with groups and partial paths. This
    mapping will be used by the seed generator and the static
    handlers.

    @method _registerGroup
    @private
    @param {Object} group
        @param {String} group.prefix
        @param {Object} group.urls
        @param {Object} group.config
    **/
    _registerGroup: function (group) {
        this._groups = this._groups || [];
        this._groups.push(group);
    },

    /**
    Given an absolutePath to a file, read the contents and return the stats.

    The `data` object return via the callback is of the form:
    <dl>
      <dt>data.body</dt>
        <dd>The `Buffer` instance which represents the contents of the file.</dd>
      <dt>data.stat</dt>
        <dd>The `stat` object for `fs.stat`.</dd>
    </dl>
    @method getAssetFromFS
    @protected
    @param {String} absolutePath 
    @param {Function} callback callback(err, data)
      @param {Error|String} err
      @param {Object} data
    **/
    getAssetFromFS: function (absolutePath, callback) {

        // Bubble up an error if the request fails to
        // normalize the path.
        if (!absolutePath) {
            return callback(new Error('File not found: ' + absolutePath));
        }

        // TODO: try/catch

        libfs.stat(absolutePath, function (err, stat) {

            if (err || !stat.isFile()) {
                return callback(new Error('File not found: ' + absolutePath));
            }

            libfs.readFile(absolutePath, function (err, data) {

                if (err) {
                    return callback(new Error('Error reading file: ' + absolutePath));
                }
                callback(null, {
                    body: data,
                    stat: stat
                });

            }); // fs.readFile

        }); // fs.stat

    },

    getContentType: function (url) {
        var ext,
            mt,
            cs;

        ext = libpath.extname(url).toLowerCase();
        // removing the . when posible
        ext = ext.indexOf('.') === 0 ? ext.slice(1) : ext;
        // computing mime type based on the extension
        mt = (ext && libmime.types[ext]) || 'text/plain';
        // computing charset based on the mime type
        cs = libmime.charsets.lookup(mt, 'UTF-8');

        return (mt + ';charset=' + cs).toLowerCase();
    },

    getGroupFromURL: function (url) {
        var groups = this._groups,
            i;

        if (url) {
            for (i in groups) {
                if (groups.hasOwnProperty(i) &&
                        (url.indexOf(groups[i].prefix) === 0)) {

                    // TODO: apply extra filters from matching group

                    return groups[i];

                }
            }
        }
    },

    /**
    Only expose a set of files within a directory while keeping the rest of
    the files way from public access.

    The `mapping` object is of the following form:
        { public-path: absolute-path-to-resource, .. }

    e.g.
    All the files in __dirname + '/htdocs/protected/' are not accessible
    publicly except for the 2 defined resources.

        app.use('/protected', statichandler.map({
            "one.html": __dirname + "/htdocs/protected/one.html",
            "assets/style.css": __dirname + "/htdocs/protected/assets/style.css"
        });

    @method map
    @public
    @param {String} groupName name of the group to associate this mapping with
    @param {Object} urls the url to absolute file mapping (see description)
    @param {Object} config
    @return {Function} express static middleware
    **/
    map: function (groupName, urls, config) {

        config = config || {};
        urls = urls || {};

        // TODO: validate groupName

        var prefix = '/' + groupName + '/',
            my = this;

        this._registerGroup({
            prefix: prefix,
            urls: urls,
            config: config
        });

        debug('serving static map thru %d', prefix);

        return function (req, res, next) {

            // valid urls are only those that matches prefix
            var url = req.url.indexOf(prefix) === 0 && req.url.slice(prefix.length),
                stream;

            if (url && urls[url]) {

                stream = new SendStream(res, config);
                stream.maxage(config.maxAge || 0);
                stream.on('error', function (err) {
                    next(err);
                });

                if ('GET' !== req.method && 'HEAD' !== req.method) {
                    return next();
                }

                my.getAssetFromFS(libpath.join(config.root || '', urls[url]), function (err, data) {
                    if (err) {
                        return stream.onStatError(err);
                    }
                    stream.pipe(data.body, data.stat);
                });

            } else {
                next();
            }

        };

    },

    /**
    Serves static modules and assets based on the rootPath.
    This method uses `express.static` under the hood, and
    configurations will be hand over to that middleware.

    @method folder
    @public
    @param {String} groupName the name of the group to associate this
    folder mapping
    @param {String} rootPath The path to the folder that should
    be exposed. This will be appended to `config.root` 
    if specified, otherwise will be used as the actual root.
    @param {Object} config the static handler configuration for
    `express.static`.

        @param {String} config.root absolute root path
        @param {Object} config.urls optional hash table with the
        public url mapping the filesystem path per file.

    @return {function} express static middlewares
    **/
    folder: function (groupName, rootPath, config) {

        // TODO: validation
        config = config || {};

        var prefix = '/' + groupName + '/',
            middleware;

        this._registerGroup({
            prefix: prefix,
            rootPath: rootPath,
            config: config
        });

        rootPath = libpath.join(config.root || '', rootPath || '');
        middleware = express['static'](rootPath, config);

        debug('serving static folder %d', rootPath);

        return function (req, res, next) {

            // valid urls are only those that matches prefix
            var url = req.url.indexOf(prefix) === 0 && req.url.slice(prefix.length);

            if (url) {

                if ('GET' !== req.method && 'HEAD' !== req.method) {
                    return next();
                }
                req.url = url;
                middleware(req, res, next);

            } else {
                next();
            }

        };

    },

    /**
    Serves combined modules and assets based on the mapping.

    @method combine
    @public
    @param {Object} config the static handler configuration
        @param {String} config.comboBase optional if not specified, default to ``
        @param {String} config.comboSep optional if not specified, default to `~`
        @param {Number} config.maxAge optional number of seconds to
        cache the combo url. By default 1 year.
        @param {Object} config.mimeTypes optional mapping between
        file extensions and response header for content-type. e.g.
        { 'html': 'text/plain' }

    @return {function} express static middlewares
    **/
    combine: function (config) {

        config = config || {};

        var comboSep  = config.comboSep || '~',
            comboBase = config.comboBase || '',
            my        = this,

            // Path used for the custom combo route
            customComboPath = '/combo/cc/p';

        function normalize(url) {

            var group = my.getGroupFromURL(url);

            if (group) {

                url = url.slice(group.prefix.length);

                if (group.urls) {
                    // we need to force mapping
                    if (group.urls[url]) {
                        return libpath.join((group.config.root || ''),
                            group.urls[url]);
                    }
                } else {
                    // we need to rely on the filesystem, no mapping
                    // is needed for this group.
                    url = libpath.join(group.config.root || '',
                            group.rootPath, url);
                    if (url.indexOf(libpath.join(group.config.root || '',
                            group.rootPath)) === 0) {
                        // If the path is controlled, which means the url is
                        // not attempting to traverse above the root path, we
                        // are good to go.
                        return url;
                    }
                }

            }

        }

        debug('enabling combo thru %d', comboBase);

        return function (req, res, next) {

            var body = [],
                urls = [],
                mtypes = config.mimeTypes || null,
                newestStat,
                pending,
                stream,
                customComboRequest;

            // If the request is a custom-encoded combo url, redirect to self
            // with a path-based combo url.
            if (req.url && req.url.indexOf(customComboPath) === 0) {
                // Create a pseudo request object with the `path` property set
                // to a path relative to the route's root path. This relative
                // path is what the decoder expects.
                // /combo/cc/p/core+3.12.0+yui.js => /core+3.12.0+yui.js
                customComboRequest = {
                    path: req.url.slice(customComboPath.length)
                };

                decoder(customComboRequest, res, function (err) {
                    if (err) {
                        res.statusCode = 400;
                        return res.end();
                    }
                    return res.redirect(
                        comboBase + res.locals.paths.join(comboSep)
                    );
                });
                return;
            }

            if (req.url && req.url.indexOf(comboBase) === 0) {
                urls = req.url.slice(comboBase.length).split(comboSep);
            } else {
                return next();
            }

            pending = urls.length;

            if ('GET' !== req.method && 'HEAD' !== req.method) {
                return next();
            }

            stream = new SendStream(res, config);
            stream.maxage(config.maxAge || 0);
            stream.on('error', function (err) {
                next(err);
            });

            if (!pending) {
                return stream.error(404, new Error('No files requested.'));
            }

            if (dedupe(urls.map(function (url) {
                    var mtype,
                        ext;
                    if (!mtypes) {
                        mtype = my.getContentType(url);
                    } else {
                        ext = libpath.extname(url).toLowerCase();
                        mtype = mtypes[ext] || DEFAULT_MIME_TYPE;
                    }
                    return mtype;
                })).length !== 1) {
                return stream.error(404, new Error('Mixing content-types in combo request.'));
            }

            urls.forEach(function (url, i) {

                my.getAssetFromFS(normalize(url), function (err, data) {

                    var blob,
                        len,
                        j;

                    if (err) {
                        return stream.onStatError(err);
                    }

                    data.stat.lastModified = new Date(data.stat.mtime);

                    // TODO: validate that this statement is correct
                    // where the newest file should win
                    if (!newestStat || newestStat.lastModified < data.stat.lastModified) {
                        newestStat = data.stat;
                    }

                    body[i]  = data.body; // storing buffers before combine them
                    pending -= 1;

                    if (pending === 0) {

                        // first pass computes total length, so we can make a
                        // buffer of the correct size
                        len = 0;
                        for (j = 0; j < body.length; j += 1) {
                            len += body[j].length;
                        }
                        newestStat.size = len;
                        blob = new Buffer(len);

                        // second pass to actually fill the buffer
                        len = 0;
                        for (j = 0; j < body.length; j += 1) {
                            body[j].copy(blob, len);
                            len += body[j].length;
                        }

                        stream.pipe(blob.toString(), newestStat);
                    }

                });
            });
        };
    }
};


examples/standalone/app.js
/*
 * Copyright (c) 2013, Yahoo! Inc.  All rights reserved.
 * Copyrights licensed under the New BSD License.
 * See the accompanying LICENSE file for terms.
 */

/*jslint node:true, nomen: true*/

'use strict';

var libpath        = require('path'),
    express        = require('express'),
    exphbs         = require('express3-handlebars'),
    statichandler  = require('../../'),
    app            = express(),
    appRootPath    = __dirname;

// setup public folders
// test urls:
// /foo/bar/one.js           => $root/htdocs/public/one.js
// /foo/bar/assets/style.css => $root/htdocs/public/assets/style.css
app.use('/foo', statichandler.folder('bar', '/public', {
    // maxAge: 123,
    root: libpath.join(appRootPath, 'htdocs')
}));

// setup "protected" files by providing a specific mapping.
// "protected" does not mean "access control", but only exposing a limited
// number of files without serving the entire directory.
// .e.g. application can use a "resolver" to get that metadata generated.
// test urls:
// /baz/qux/another/two.js => $root/htdocs/protected/two.js
// /baz/qux/also/style.css => $root/htdocs/protected/assets/style.css
app.use('/baz', statichandler.map('qux', {
    "another/two.js": libpath.join('protected', 'two.js'),
    "also/style.css": libpath.join('protected', 'assets', 'style.css')
}, {
    // maxAge: 123,
    root: libpath.join(appRootPath, 'htdocs')
}));

app.use(statichandler.combine({
    comboBase: '/combo~',
    comboSep: '~'
    // maxAge: 123
}));

// template engine
app.engine('handlebars', exphbs());
app.set('view engine', 'handlebars');

// creating a page with YUI embeded
app.get('/', function (req, res, next) {
    res.render('page');
});

// listening
app.set('port', process.env.PORT || 8666);
app.listen(app.get('port'), function () {
    console.log("Server listening on port " +
        app.get('port') + " in " + app.get('env') + " mode");
});


lib/send.js
/*jslint node:true, nomen: true */

/**
The `send` module was inspired by the `Send` component shipped
with Connect's middleware, and it provides a basic api to stream
static assets from a blob. Probably 80% of the code comes
from the original implementation ;) but with better format!

@module send
**/

'use strict';

var debug = require('debug')('express-combo:send'),
    parseRange = require('range-parser'),
    StringStream = require('./string'),
    mime = require('mime'),
    fresh = require('fresh'),
    http = require('http');


/**
 * Return an ETag in the form of `"<size>-<mtime>"`
 * from the given `stat`.
 *
 * @param {Object} stat
 * @return {String}
 * @private
 */
function etag(stat) {
    return '"' + stat.size + '-' + Number(stat.mtime) + '"';
}

/**
 * decodeURIComponent.
 *
 * Allows V8 to only deoptimize this fn instead of all
 * of send().
 *
 * @param {String} path
 * @private
 */
function decode(path) {
    try {
        return decodeURIComponent(path);
    } catch (err) {
        return -1;
    }
}

/**
Initialize a `SendStream` with the given `data`.

Events:

 - `error` an error occurred
 - `stream` file streaming has started
 - `end` streaming has completed

@class SendStream
@static
@uses TODO
@constructor
@param {Response} res
@param {Object} options
@protected
**/
function SendStream(res, options) {
    this.res = res;
    this.req = res.req;
    this.path = res.req.originalUrl;
    this.options = options || {};
    this.maxage(0);
}

// Inherits from `Stream`.
require('util').inherits(SendStream, require('stream'));

/**
Set max-age to `ms`.

@param {Number} ms
@return {SendStream}
**/
SendStream.prototype.maxage = function (ms) {
    if (Infinity === ms) {
        ms = 60 * 60 * 24 * 365 * 1000;
    }
    debug('max-age %d', ms);
    this._maxage = ms;
    return this;
};
/**
Emit error with `status`.

@param {Number} status
@protected
**/
SendStream.prototype.error = function (status, err) {
    var res = this.res,
        msg = http.STATUS_CODES[status];

    err = err || new Error(msg);
    err.status = status;
    if (this.listeners('error').length) {
        return this.emit('error', err);
    }
    res.statusCode = err.status;
    res.end(msg);
};

/**
Check if this is a conditional GET request.

@protected
@return {Boolean}
**/
SendStream.prototype.isConditionalGET = function () {
    return this.req.headers['if-none-match'] ||
        this.req.headers['if-modified-since'];
};

/**
Strip content-* header fields.

@protected
**/
SendStream.prototype.removeContentHeaderFields = function () {
    var res = this.res;
    Object.keys(res._headers).forEach(function (field) {
        if (0 === field.indexOf('content')) {
            res.removeHeader(field);
        }
    });
};

/**
Respond with 304 not modified.

@protected
**/
SendStream.prototype.notModified = function () {
    var res = this.res;
    debug('not modified');
    this.removeContentHeaderFields();
    res.statusCode = 304;
    res.end();
};

/**
Check if the request is cacheable, aka
responded with 2xx or 304 (see RFC 2616 section 14.2{5,6}).

@protected
@return {Boolean}
**/
SendStream.prototype.isCachable = function () {
    var res = this.res;
    return (res.statusCode >= 200 && res.statusCode < 300) ||
        304 === res.statusCode;
};

/**
Handle stat() error.

@param {Error} err
@protected
**/
SendStream.prototype.onStatError = function (err) {
    var notfound = ['ENOENT', 'ENAMETOOLONG', 'ENOTDIR'];
    if (-1 !== notfound.indexOf(err.code)) {
        return this.error(404, err);
    }
    this.error(500, err);
};

/**
Check if the cache is fresh.

@return {Boolean}
@protected
**/
SendStream.prototype.isFresh = function () {
    return fresh(this.req.headers, this.res._headers);
};

/**
Pipe to `res.

@param {String} data
@param {Object} stat
@return {Stream} res
**/
SendStream.prototype.pipe = function (data, stat) {
    var self = this,
        path = this.path;

    // invalid request uri
    path = decode(path);
    if (-1 === path) {
        return this.error(400);
    }

    // null byte(s)
    if (-1 !== path.indexOf('\u0000')) {
        return this.error(400);
    }

    self.send(data, stat);

    return this.res;
};

/**
Transfer `path`.

@param {String} path
**/
SendStream.prototype.send = function (data, stat) {
    var options = this.options,
        len = stat.size,
        res = this.res,
        req = this.req,
        path = this.path,
        ranges = req.headers.range,
        offset = options.start || 0,
        bytes;

    // set header fields
    this.setHeader(stat);

    // set content-type
    this.type(path);

    // conditional GET support
    if (this.isConditionalGET() && this.isCachable() && this.isFresh()) {
        return this.notModified();
    }

    // adjust len to start/end options
    len = Math.max(0, len - offset);
    if (options.end !== undefined) {
        bytes = options.end - offset + 1;
        if (len > bytes) {
            len = bytes;
        }
    }

    // Range support
    if (ranges) {
        ranges = parseRange(len, ranges);

        // unsatisfiable
        if (-1 === ranges) {
            res.setHeader('Content-Range', 'bytes**/' + stat.size);
            return this.error(416);
        }

        // valid (syntactically invalid ranges are treated as a regular response)
        if (-2 !== ranges) {
            options.start = offset + ranges[0].start;
            options.end = offset + ranges[0].end;

            // Content-Range
            res.statusCode = 206;
            res.setHeader('Content-Range', 'bytes ' + ranges[0].start + '-' + ranges[0].end + '/' + len);
            len = options.end - options.start + 1;
        }
    }

    // content-length
    res.setHeader('Content-Length', len);

    // HEAD support
    if ('HEAD' === req.method) {
        return res.end();
    }

    this.stream(data, options);
};

/**
Stream `data` to the response.

@param {String} data the content to flush in the response
@param {Object} options
@protected
**/
SendStream.prototype.stream = function (data) {
    // TODO: this is all lame, refactor meeee
    var self = this,
        res = this.res,
        req = this.req,
        // pipe
        stream = new StringStream(data);

    this.emit('stream', stream);
    stream.pipe(res);

    // socket closed, done with the fd
    req.on('close', stream.destroy.bind(stream));

    // error handling code-smell
    stream.on('error', function (err) {
        // no hope in responding
        if (res._header) {
            req.destroy();
            return;
        }
        // 500
        err.status = 500;
        self.emit('error', err);
    });

    // end
    stream.on('end', function () {
        self.emit('end');
    });

    stream.resume();
};
/**
Set content-type based on `path`
if it hasn't been explicitly set.

@param {String} path
@protected
**/
SendStream.prototype.type = function (path) {
    var res = this.res,
        type,
        charset;
    if (res.getHeader('Content-Type')) {
        // do not overrule any previous content-type
        return;
    }
    type = mime.lookup(path);
    charset = mime.charsets.lookup(type);
    debug('content-type %s', type);
    res.setHeader('Content-Type', type + (charset ? '; charset=' + charset : ''));
};

/**
Set response header fields, most
fields may be pre-defined.

@param {Object} stat
@protected
**/
SendStream.prototype.setHeader = function (stat) {
    var res = this.res;
    if (!res.getHeader('Accept-Ranges')) {
        res.setHeader('Accept-Ranges', 'bytes');
    }
    if (!res.getHeader('ETag')) {
        res.setHeader('ETag', etag(stat));
    }
    if (!res.getHeader('Date')) {
        res.setHeader('Date', new Date().toUTCString());
    }
    if (!res.getHeader('Cache-Control')) {
        res.setHeader('Cache-Control', 'public, max-age=' + (this._maxage / 1000));
    }
    if (!res.getHeader('Last-Modified')) {
        res.setHeader('Last-Modified', stat.mtime.toUTCString());
    }
};


module.exports = SendStream;
